### DISCUSSION TITLE
Issues with TS in dev mode after upgrade to vite-rails 3, vite 2.5.1

### DISCUSSION THREAD

**Discussion by keithlayne** [ElMassimo/vite_ruby#122] [Q&A] [answered]
I tried updating my app yesterday to latest of `vite_ruby`, `vite_rails`, `vite`, and `vite-plugin-ruby` and ran into problems. There are several tools in play here, and I'm just throwing this out there in case anyone has seen similar.

What appears to be happening is that the transpile output is not removing some or all of type imports, which causes a `module doesn't export X` error.  I _think_ that any `import type` lines are removed, and I _think_ the rest of the types are stripped. I'm going from memory here because I pretty much immediately rolled back because I had work to do.

I realize that `vite` and `esbuild` are the more likely culprits here, but if anyone else has seen similar issues after an upgrade and maybe spent more time debugging that would be great.

I can probably do a little more digging but it will have to wait at least a day.

**Comment by ElMassimo** [comment]
Hi Keith, sorry to hear that.

I've been using TypeScript in all recent projects, but I haven't faced a problem like that where esbuild does not remove TypeScript syntax. Do the files that fail use the standard extension `.ts`?

Perhaps you can share your `vite.config.ts` file, in case it's related to configuration or the plugins that are being applied.

**Comment by keithlayne** [comment]
I created a branch this morning for debugging, here are the steps I've taken:

* Update vite_rails to `~> 3.0`
* run `bin/vite upgrade` which gets new versions:
  * esbuild@0.12.24
  * vite-plugin-ruby@3.0.1
  * vite@2.5.1
* run `bin/vite clobber`
* run `bin/rails tmp:clear'

AFAICT this gets me a clean, upgraded setup, and then I see the behavior.

I confirm that import lines like:

```ts
import { Foo, FooProps } from './foo';
```
are not transformed with this setup.

```ts
import type { FooProps } from './foo';
```
is however removed from the output.

I suspect a vite/esbuild bug at this point. I'll look through their issues again.

**Comment by keithlayne** [comment, marked as answer]
There's a similar (old) esbuild issue here: https://github.com/evanw/esbuild/issues/341

I'm already using `isolatedModules`, which doesn't seem relevant, as I'm seeing direct type exports hit this issue, with no type-reexports involved. Could probably work around this by separating type and value imports, but I'm not going to do that.

Ah, just found the issue, finally: https://github.com/evanw/esbuild/issues/1525

So, definitively an esbuild bug with `"importsNotUsedAsValues": "error"` (or `"preserve"` I guess).
